zebra/kernel_netlink.c

```diff
@@ -1080,6 +1132,288 @@ int netlink_request(struct nlsock *nl, void *req)
 	return 0;
 }
 
+static int nl_batch_read_resp(struct nl_batch *bth)
+{
+	struct nlmsghdr *h;
+	struct sockaddr_nl snl;
+	struct msghdr msg;
+	int status;
+	const struct nlsock *nl;
+	struct nl_msg_batch_item *item, *from_item;
+
+	nl = &(bth->zns->nls);
+
+	msg.msg_name = (void *)&snl;
+	msg.msg_namelen = sizeof(snl);
+
+	from_item = nl_batch_list_first(&(bth->items));
+
+	status = netlink_recv_msg(nl, msg, nl_batch_rx_buf,
+				  sizeof(nl_batch_rx_buf));
+	if (status == -1 || status == 0)
+		return status;
+
+	for (h = (struct nlmsghdr *)nl_batch_rx_buf;
+	     (status >= 0 && NLMSG_OK(h, (unsigned int)status));
+	     h = NLMSG_NEXT(h, status)) {
+
+		/*
+		 * Find the corresponding batch item. Received responses are in
+		 * the same order as requests we sent, so we can simply at same
+		 * time iterate over the batch item list and match responses
+		 * with requests.
+		 */
+		frr_each_from(nl_batch_list, &(bth->items), item, from_item) {
```

Comment:
According to the docs the `frr_each_from` can resume iteration after breaking out of the loop by keeping the from value persistent and reusing it for the next loop. And that's what I wanted to do here. Since both `nl_batch_list` and Netlink response contains messages sorted by their sequence numbers, so we can match items using two pointers technique.

Reply:
ah, sure, sorry, I didn't read the 'from' macro well enough!